src/jsxc.lib.httpupload.js

```diff
@@ -0,0 +1,167 @@
+         attachment.addClass('jsxc_' + file.type.replace(/^([^/]+)\/.*/, '$1'));
+
+         msg.empty().append(attachment);
+
+         if (FileReader && file.type.match(/^image\//)) {
+            var img = $('<img alt="preview">').attr('title', file.name);
+            img.attr('src', jsxc.options.get('root') + '/img/loading.gif');
+            img.appendTo(attachment);
+
+            var reader = new FileReader();
+
+            reader.onload = function() {
+               img.attr('src', reader.result);
+            };
+
+            reader.readAsDataURL(file);
+         } else {
+            attachment.text(file.name + ' (' + file.size + ' byte)');
+         }
+
+         $('<button>').addClass('jsxc_btn jsxc_btn-primary').text($.t('Send')).click(function() {
+            jsxc.httpupload._sendFile(jid, bid, img, file);
+
+            jsxc.gui.window.hideOverlay(bid);
+
+            msg.remove();
+
+         }).appendTo(msg);
+
+         $('<button>').addClass('jsxc_btn jsxc_btn-default').text($.t('Abort')).click(function() {
+            jsxc.gui.window.hideOverlay(bid);
+         }).appendTo(msg);
+      });
+   },
+
+   _sendFile: function(jid, bid, img, file){
+      var uploadUrl = jsxc.options.get('httpupload').server;
+      var iq = $iq({
+         to: uploadUrl,
+         type: 'get'
+      }).c('request', {
+         xmlns: jsxc.httpupload.CONST.HTTPUPLOAD
+      }).c('filename').t(file.name).up().c('size').t(file.size);
+
+      jsxc.xmpp.conn.sendIQ(iq, function(resIq) {
+         var slot = resIq.getElementsByTagName("slot");
+         if(slot.length > 0)
+         {
+            var putUrl = slot[0].getElementsByTagName("put")[0].firstChild.nodeValue;
+            $.ajax({
+               url:putUrl,
+               type:'PUT',
+               contentType: 'application/octet-stream',
+               data:file,
+               crossDomain: true,
+               processData:false,
+               complete: function(xhr) {
+                  if (xhr.readyState === 4) {
+                     if (xhr.status === 201) {
+                        var uid = Date.parse(new Date()) + ':msg';
+                        var message = jsxc.gui.window.postMessage({
+                           _uid: uid,
+                           bid: bid,
+                           direction: 'out',
+                           dataType: 'text',
+                           attachment: {
+                              name: file.name,
+                              size: file.size,
+                              type: file.type,
+                              data: (file.type.match(/^image\//)) ? img.attr('src') : null
+                           }
+                        });
+                        jsxc.gui.window.updateProgress(message, file.size, file.size);
+                        jsxc.xmpp.sendMessage(bid, putUrl, uid);
+                     }
+                  } else {
+                     jsxc.warn('ajax put failed');
```

Comment:
This message should be a little more detailed and specially catch the error conditions from http://xmpp.org/extensions/xep-0363.html#errors